### PR TITLE
refactor: migrate to viper configure manage

### PR DESCRIPTION
**Pull request by bzp2010**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [ ] New feature provided
- [x] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

The configuration management is implemented by `viper` library instead of reading yaml file directly.
It will provide support for YAML, JSON, TOML and other configuration file formats, and even remote loading through etcd, which will be controlled by `-c` or `--config` parameters.

And no incompatible updates introduced.

**Related issues**

part of #1931

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

**Comment by bzp2010** [comment]
#### 2021-06-22 Update1
There are some key problems in refactoring.

1. some functions of configuration file path and `work-dir` parameter overlap
2. ci test script execution path problem causes the program unable to automatically find the configuration file location

File: api/internal/conf/conf.go [hunk 1]
```diff
@@ -19,14 +19,13 @@ package conf
 import (
 	"fmt"
 	"io/ioutil"
-	"log"
 	"os"
 	"path/filepath"
 	"runtime"
 	"strings"
 
+	"github.com/spf13/viper"
```
Comment (tokers) [inline api/internal/conf/conf.go:27]: IMHO Why not just use `cobra`.

Comment (bzp2010) [reply api/internal/conf/conf.go:27]: `cobra` only provides command line management, 
`viper` provides configuration file reading and parsing function.

File: api/go.mod [hunk 2]
```diff
@@ -19,32 +18,27 @@ require (
...
-	github.com/json-iterator/go v1.1.11 // indirect
 	github.com/prometheus/client_golang v1.8.0 // indirect
 	github.com/satori/go.uuid v1.2.0
 	github.com/shiningrush/droplet v0.2.6-0.20210127040147-53817015cd1b
 	github.com/shiningrush/droplet/wrapper/gin v0.2.1
 	github.com/sirupsen/logrus v1.7.0 // indirect
 	github.com/sony/sonyflake v1.0.0
 	github.com/spf13/cobra v0.0.3
-	github.com/spf13/pflag v1.0.5 // indirect
+	github.com/spf13/viper v1.8.1
```
Comment (nic-chen) [inline api/go.mod:32]: I remember that the license of a dependency of `viper` was not compatible with Apache 2.0,  please confirm.

Comment (bzp2010) [reply api/go.mod:32]: Hi, @nic-chen. I'm sorry for the slow response.

I used `google/go-licenses` to generate a list of the `spf13/viper`'s dependencies and licenses.

| library | URL | license |
|  ----  | ----  | ---- |
| github.com/mitchellh/mapstructure | https://github.com/mitchellh/mapstructure/blob/master/LICENSE | MIT |
| github.com/magiconair/properties | https://github.com/magiconair/properties/blob/master/LICENSE.md | BSD-2-Clause |
| github.com/spf13/afero | https://github.com/spf13/afero/blob/master/LICENSE.txt | Apache-2.0 |
| github.com/spf13/pflag | https://github.com/spf13/pflag/blob/master/LICENSE | BSD-3-Clause |
| github.com/hashicorp/hcl | https://github.com/hashicorp/hcl/blob/master/LICENSE | MPL-2.0 |
| github.com/pelletier/go-toml | https://github.com/pelletier/go-toml/blob/master/LICENSE | Apache-2.0 |
| github.com/spf13/viper | https://github.com/spf13/viper/blob/master/LICENSE | MIT |
| github.com/spf13/cast | https://github.com/spf13/cast/blob/master/LICENSE | MIT |
| github.com/spf13/jwalterweatherman | https://github.com/spf13/jwalterweatherman/blob/master/LICENSE | MIT |
| github.com/subosito/gotenv | https://github.com/subosito/gotenv/blob/master/LICENSE | MIT |
| github.com/fsnotify/fsnotify | https://github.com/fsnotify/fsnotify/blob/master/LICENSE | BSD-3-Clause |
| gopkg.in/yaml.v2 | https://github.com/go-yaml/yaml | Apahce-2.0 |
| golang.org/x/text | https://golang.org/x/text | BSD-3-Clause |
| gopkg.in/ini.v1 | https://gopkg.in/ini.v1 | Apache-2.0 |

And these licenses are also used in current Manager API dependencies.

File: api/internal/conf/conf.go (hunk 1 above)
Comment (tokers) [reply api/internal/conf/conf.go:27]: Got it.

**Review by tokers — APPROVED** [review]

**Review by starsz — APPROVED** [review]

File: api/go.mod (hunk 2 above)
Comment (nic-chen) [reply api/go.mod:32]: I'm not sure about `MPL-2.0`, @moonming  please have a look, thanks.

Comment (bzp2010) [reply api/go.mod:32]: ### Description
`Viper` library uses the API of `hashicorp` open source repository using `MPL-2.0` license, but does not use its code directly.

### Solution
If importing a library using `MPL-2.0` license is not feasible, I will consider implementing a shim code and replacing it with `go mod replace`. It emulates the API of that library and provides empty results.

It's like the following image.
image

So is such a solutionfeasible?

Comment (nic-chen) [reply api/go.mod:32]: I searched the go project under apache, and many projects also use viper
for example: 

https://github.com/apache/skywalking-banyandb

https://github.com/apache/camel-k

I have no problem here.

Comment (moonming) [reply api/go.mod:32]: MPL-2.0 maybe can NOT in apache projects: https://www.apache.org/legal/resolved.html

**Review by juzhiyuan — CHANGES_REQUESTED** [review]
see https://github.com/apache/apisix-dashboard/pull/1946#discussion_r676365553

File: api/go.mod (hunk 2 above)
Comment (bzp2010) [reply api/go.mod:32]: @moonming Sorry to bother you again.

- I have read this specification [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html) and made a query in JIRA, I found the following discussion, [LEGAL-562: Handling dependency licenses for the Solr Operator, a go project](https://issues.apache.org/jira/browse/LEGAL-562).
I think our situation is similar to that in the article. We do not include the code under the MPL-2.0 license in the code distribution. The code is downloaded by the user during the compilation process. In the docker image involving binary distribution, we should modify the license and note files and include them in the image as a reminder. Just like these project, [apache/skywalking-banyandb](https://github.com/apache/skywalking-banyandb).

- If we really can't include the dependencies of this library, can the above scheme of replacing the dependent library code with shim? [here](https://github.com/apache/apisix-dashboard/pull/1946#discussion_r670114961)

🤔 What do you think of this? Thx.

Comment (bzp2010) [reply api/go.mod:32]: ping @moonming Any changes here?

Comment (moonming) [reply api/go.mod:32]: Having a dependency with a Category B license is fine

**Review by moonming — APPROVED** [review]

**Review by juzhiyuan — APPROVED** [review]

**Pull request merged**